toughday/src/main/java/com/adobe/qe/toughday/internal/core/distributedtd/redistribution/TaskBalancer.java

```diff
@@ -0,0 +1,287 @@
+import com.adobe.qe.toughday.internal.core.distributedtd.splitters.PhaseSplitter;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.http.HttpResponse;
+import org.apache.http.util.EntityUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.*;
+
+import static com.adobe.qe.toughday.internal.core.distributedtd.HttpUtils.HTTP_REQUEST_RETRIES;
+
+/**
+ * Class responsible for balancing the work between the agents running the cluster
+ * whenever the number of agents is changing.
+ */
+public class TaskBalancer {
+    protected static final Logger LOG = LogManager.getLogger(Engine.class);
+    private static TaskBalancer instance = null;
+
+    private RedistributionStatus status = RedistributionStatus.UNNECESSARY;
+    private final PhaseSplitter phaseSplitter = new PhaseSplitter();
+    private final HttpUtils httpUtils = new HttpUtils();
+    private final ConcurrentLinkedQueue<String> inactiveAgents = new ConcurrentLinkedQueue<>();
+    private final ConcurrentLinkedQueue<String> recentlyAddedAgents = new ConcurrentLinkedQueue<>();
+    private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();
+
+    public enum RedistributionStatus {
+        UNNECESSARY,            // no need to redistribute the work
+        SCHEDULED,              // redistribution is scheduled
+        RESCHEDULE_REQUIRED,   /* nr of agents running in the cluster has changed while the work was redistributed so
+                                 * we must reschedule the process */
+        EXECUTING               // work redistribution process is executing
+    }
+
+    /**
+     * Returns a list with all the agents that failed to respond to the heartbeat request sent
+     * by the driver.
+     */
+    public ConcurrentLinkedQueue<String> getInactiveAgents() {
+        return this.inactiveAgents;
+    }
+
+    private TaskBalancer() {}
+
+    /**
+     * Returns a singleton instance of this class.
+     */
+    public static TaskBalancer getInstance() {
+        if (instance == null) {
+            instance = new TaskBalancer();
+        }
+
+        return instance;
+    }
+
+    private Map<String, Long> getTestSuitePropertiesToRedistribute(TestSuite taskTestSuite) {
+        HashMap<String, Long> map = new HashMap<>();
+        taskTestSuite.getTests().forEach(test -> map.put(test.getName(), test.getCount()));
+
+        return map;
+    }
+
+    private void addNewAgent(String agentIdentifier) {
+        this.recentlyAddedAgents.add(agentIdentifier);
+    }
+
+    private void addInactiveAgent(String agentIdentifier) {
+        this.inactiveAgents.add(agentIdentifier);
+    }
+
+    private void waitUntilAllAgentsAreReadyForRebalancing(Queue<String> agents) {
```

Comment:
The assumption is that when an Agent receives a redistribution request, the Configuration object is already created so that we will be able to modify the properties of certain objects (RunMode, TestSuite). The following will explain this concern better:

1. A new agent(10.0.0.1) joins the cluster.
2. The work redistribution process begins => agent 10.0.0.1 receives a Configuration in YAML format
3. A new agent(10.0.0.2) joins the cluster
4. The work redistribution process begins => agent 10.0.0.1 receives instructions to modify certain properties of the run mode.

If agent (10.0.0.1) failed to create the Configuration before step 4, it will not be able to process the redistribution request.